common/reactive/src/main/java/io/helidon/common/reactive/MultiConcatArray.java

```diff
@@ -70,38 +141,215 @@ public void onNext(T item) {
 
         @Override
         public void onError(Throwable throwable) {
+            REQUESTED.setOpaque(this, CANCEL);
             downstream.onError(throwable);
         }
 
         @Override
         public void onComplete() {
-            long produced = this.produced;
-            if (produced != 0L) {
-                this.produced = 0L;
-                super.produced(produced);
+            Thread current = Thread.currentThread();
+            if (LASTTHREADCOMPLETING.getOpaque(this) == current) {
+                redo = true;
+                return;
+            }
+
+            LASTTHREADCOMPLETING.setOpaque(this, current);
+            VarHandle.storeStoreFence();
```

Comment:
Reading this again, strictly speaking this fence is not needed, because we have stronger fences between the stores targeted by this fence. But no need to change the code as the cost of this fence on the target platforms is zero.